Hutchpp

joanrue opened the pull request:
Add trace method to LinOp (hutch++ estimation)
Fixed hutch++ algorithm bug (wrong random distribution)
Improve documentation

joanrue committed:
[math/linalg] Add hutchpp

Added hutch-plusplus algorithm for stochastic trace estimation

joanrue committed:
[abc/operator] Add 'fro' algorithm for lipschitz computation (hutch++)

joanrue committed:
[math/linalg] DOC: add gpu arg in hutch++ docstring

joanrue committed:
[hutch++][BugFix] use binomial dist. and default_rng

joanrue committed:
[LinOp.trace] add LinOp.trace(), use it for lipschitz computation

joanrue committed:
Merge branch 'v2-dev' into hutchpp

joanrue committed:
Merge branch 'v2-dev' into hutchpp

joanrue committed:
[hutch++][BugFix] Delete 'gpu' key from kwargs

joanrue committed:
[hutch++] Change default to high precision

The default trace estimation is now performed at high precision (4002 queries), whereas the lipschitz calculation is performed at low precision (126 queries)

joanrue committed:
[LinOp/lipschitz/fro][Doc update]

Added more information on the behavior of the frobenius norm as an estimate for the lispchitz cst

joanrue commented:
@SepandKashani There is an issue with this pull-request related to a wrongdoing in my last re-basing. The commits are duplicated from the merging of your previous push to v2-dev.

SepandKashani commented:
@joanrue I will have a look in a few hours.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -6,6 +6,7 @@
 import numpy as np
 import scipy.sparse.linalg as splin
 
+import pycsou.math.linalg as pycl
```
This will create a circular dependency since in  ``pycsou.math.linalg`` you are importing ``pycsou.abc.operator``.
Make this import from within the scopes of the functions/methods that use the routines in linalg.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1769,19 +1770,53 @@ def lipschitz(self, recompute: bool = False, algo: str = "svds", gpu: bool = Fal
         """
         if algo == "fro":
-            raise NotImplementedError
+            kwargs.update(dict(m=126, gpu=gpu))
```
I think you could leave the user the possibility to set ``m`` as an optional keyword argument and set it to 126 only if it is not provided, i.e.

```python
if 'm' in kwargs.keys():
     kwargs.update(dict(gpu=gpu))
else:
    kwargs.update(dict(m=126, gpu=gpu))
```
or something along those lines.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1769,19 +1770,53 @@ def lipschitz(self, recompute: bool = False, algo: str = "svds", gpu: bool = Fal
-        .. todo::
-            Add support for trace-based estimate (@Joan).
         """
         if algo == "fro":
```
The lipschitz constant should not be recomputed if it was already computed (e.g. `self._lipschitz < np.infty`) and if the user hasn't explicitely set `recompute` to `True`. Otherwise each time we call the method it will recompute from scratch instead of giving us the value stored in `_lipschitz`. I therefore recommend the following structure:

```python3
 if recompute or (self._lipschitz == np.infty):
            if algo == "fro":
                        if np.diff(self.shape) > 0:
                                    self._lipschitz = np.sqrt(self.cogram().trace(**kwargs))
                        else:
                                    self._lipschitz = np.sqrt(self.gram().trace(**kwargs))
            elif algo == "svds":
                        kwargs.update(dict(k=1, which="LM", gpu=gpu))
                        self._lipschitz = self.svdvals(**kwargs)
            else:
                        raise NoteImplementedError
return self._lipschitz
```

This means that by default, no computation is done and the method simply returns the value of the attribute `_lipschitz` unless the ``self._lipschitz == np.infty`` or ``recompute=True``.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1769,19 +1770,53 @@ def lipschitz(self, recompute: bool = False, algo: str = "svds", gpu: bool = Fal
+            kwargs.update(dict(xp=xp))
+            del kwargs["gpu"]
+
+        return pycl.hutchpp(self, **kwargs)
```
you should raise a `NotImplementedError` if the operator is not square, i.e. `self.shape[0] != self.shape[1]`.

matthieumeo commented on src/pycsou/math/linalg.py:
```diff
@@ -0,0 +1,84 @@
+@pycrt.enforce_precision(o=True)
+def hutchpp(linop: pyco.LinOp, m: int = 4002, xp: pyct.ArrayModule = np, seed: float = 0):
+    r"""
+    Computes an stocastic estimate of the trace for linear operators based on the Hutch++ algorithm (specifically,
```
Typo: "an stocastic" -> "a stochastic"

matthieumeo commented on src/pycsou/math/linalg.py:
```diff
@@ -0,0 +1,84 @@
+def hutchpp(linop: pyco.LinOp, m: int = 4002, xp: pyct.ArrayModule = np, seed: float = 0):
+    r"""
+    Computes an stocastic estimate of the trace for linear operators based on the Hutch++ algorithm (specifically,
+    algorithm 3 of the paper https://arxiv.org/abs/2010.09649.
```
missing parenthesis.

matthieumeo replied:
For hyperlinks in reStructuredText, use  \`this paper <https://arxiv.org/abs/2010.09649>\`_

matthieumeo commented on src/pycsou/math/linalg.py:
```diff
@@ -0,0 +1,84 @@
+        based on the analysis of variance described in theorem 10 of the Hutch++ paper. This default number of queries
+        corresponds to having an estimation error smaller than 0.01 with a probability of 0.9.
+    xp:  pycsou.util.ptype.ArrayModule
+         Which array module to use to represent the output.
```
Which array module to use for the computation.

matthieumeo commented on src/pycsou/math/linalg.py:
```diff
@@ -0,0 +1,84 @@
+    xp:  pycsou.util.ptype.ArrayModule
+         Which array module to use to represent the output.
+    seed: float
+        Seed for the random number generator.
```
I think `seed` are integers but not sure...

matthieumeo commented on src/pycsou/math/linalg.py:
```diff
@@ -0,0 +1,84 @@
+
+    Notes
+    -----
+    This function calls Scipy’s function: :py:func:`scipy.sparse.linalg.qr`. See the documentation of this function
```
Remove I suggest we use the numpy qr function so that cupy and numpy have the same signature.

matthieumeo commented on src/pycsou/math/linalg.py:
```diff
@@ -0,0 +1,84 @@
+    for more information on its behaviour and the underlying LAPACK routines it relies on.
+    """
+
+    if xp == np:
```
I would go for something simpler: 

```python 
import dask.array as da

xlin=xp.linalg
if xp == da:
   kwargs = {}
else:
   kwargs = {"mode": "reduced"}
```

matthieumeo commented on src/pycsou/math/linalg.py:
```diff
@@ -0,0 +1,84 @@
+            "of queries is larger or equal to the dimension of the linear operator.",
+            UserWarning,
+        )
+        return xp.trace(linop.apply(xp.eye(d)))
```
You could store the diagonal only instead of the entire matrix: 

```python
return xp.sum(xp.array([linop.apply(e)[i] for i, e in enumerate(xp.eye(d)]))
```

matthieumeo commented on src/pycsou/math/linalg.py:
```diff
@@ -0,0 +1,84 @@
+    rng = xp.random.default_rng(seed=seed)
+    s = rng.standard_normal(size=(d, (m + 2) // 4))
+    g = rng.binomial(n=1, p=0.5, size=(d, (m - 2) // 2)) * 2 - 1
+    q, _ = xlin.qr(linop.apply(s.T).T, **kwargs)
```
Careful, with dask the `qr` function only works for tall-and-skinny or  short-and-fat arrays. This means you have to rechunck the output before calling `qr` with dask. See the `NotImplementedError` text in https://docs.dask.org/en/latest/_modules/dask/array/linalg.html#qr

matthieumeo replied:
In our case you'll want to go for the tall-and-skinny variant.

matthieumeo reviewed (changes requested):
Very nice Joan, thanks a lot. Some points must be addressed: make sure that the lipschitz constant is not recomputed unnecessarily, raise an error if trace is called on a non-square operator, simplify the module-agnostic imports in `hutchpp` and handle the case of Dask which can only compute QR decompositions on tall-and-skinny arrays. See my comments for more details.

joanrue committed:
[linop/lipschitz][BugFix] circular dependencies

joanrue committed:
[linop/lipschitz][BugFix] check recomputing conidtions before recomputing lipschitz

joanrue committed:
[linop/lipschitz][DOC] describe m param in doc

joanrue committed:
[linalg/hutch++][DOC] fix typos

joanrue committed:
[linalg/hutch++][DOC] add example stochastic trace estimation

joanrue committed:
[linalg/hutch++][ENH] clean array mod agnosticity

joanrue committed:
[linalg/hutch++][ENH] compute trace more efficiently memorywise

joanrue committed:
[linalg/hutch++][ENH] add support for dask random generator and qr func

joanrue committed:
[linop/lipschitz/fro] return array with shape (1,), consistency with svds

joanrue committed:
[linop/lipschitz/fro][DOC] add example comparing svds, trace, and hutch++

joanrue committed:
[operator/trace] Move 'trace' method from LinOp class to SquareOp class

joanrue committed:
[operator/LinOp] specialize (co)gram as SquareOp
This will allow accessing the SquareOp method .trace() when computing the Lipschitz

joanrue committed:
[operator/SquareOp][DOC] add few words on trace method

joanrue committed:
[operator/LinOp][ENH] LinOp automatically specializes to SquareOp

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1632,6 +1633,8 @@ def __init__(self, shape: pyct.NonAgnosticShape):
         super(LinOp, self).__init__(shape)
         self._diff_lipschitz = 0
+        if shape[0] == shape[1]:
+            self.__class__ = SquareOp
```
Does this really work ? I'm having second thoughts about this automatic recasting of ``LinOps`` with identical domain/co-domain dimensions... It's probably not very recommended. The cleanest would probably be to keep the trace method in ``SquareOp`` and let the user instantiate from the right class if he wants to benefit from the ``trace`` method. @joanrue, @SepandKashani some thoughts?

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1752,11 +1755,11 @@ def lipschitz(self, recompute: bool = False, algo: str = "svds", gpu: bool = Fal
             Optional arguments to be passed to the algorithm used for computing the Lipschitz constant.
             The parameter ``tol`` of Scipy's :py:func:`scipy.sparse.linalg.svds` routine can be particularly interesting to reduce
-            the compute time of the Lipschitz constant.
+            the compute time of the Lipschitz constant. The parameter ``m`` of LinOps :py:func:`pycsou.abc.operator.LinOp.trace` can also be interesting to control the number of queries used for the stochastic estimation.
```
Make it clearer which algorithm is affected by the optional keywords``tol`` or ``m``.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1769,17 +1772,56 @@ def lipschitz(self, recompute: bool = False, algo: str = "svds", gpu: bool = Fal
+                        else:
+                            import numpy as xp
+                        kwargs.update(dict(xp=xp))
+                    del kwargs["gpu"]
```
Use ``kwargs.pop("gpu")`` rather (more Pythonic).

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1897,7 +1939,7 @@ def cogram(self) -> "LinOp":
         way to compute the co-Gram operator. If the co-Gram can be computed more efficiently (e.g. with a convolution), the user should re-define this method.
         """
-        return self * self.T
+        return (self * self.T).specialize(SquareOp)
```
The output should actually be of type ``SelfAdjointOp``.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1881,9 +1923,9 @@ def gram(self) -> "LinOp":
+        return (self.T * self).specialize(SquareOp)
 
-    def cogram(self) -> "LinOp":
+    def cogram(self) -> "SquareOp":
```
The output should actually be of type ``SelfAdjointOp``.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1881,9 +1923,9 @@ def gram(self) -> "LinOp":
         way to compute the Gram operator. If the Gram can be computed more efficiently (e.g. with a convolution), the user should re-define this method.
         """
-        return self.T * self
+        return (self.T * self).specialize(SquareOp)
```
The output should actually be of type ``SelfAdjointOp``.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1867,7 +1909,7 @@ def __array__(self, dtype: typ.Optional[type] = None) -> np.ndarray:
         return self.asarray(xp=np, dtype=dtype)
 
-    def gram(self) -> "LinOp":
+    def gram(self) -> "SquareOp":
```
The output should actually be of type ``SelfAdjointOp`` (subclass of ``SquareOp``). Also don't forget to update the output type in the docstring.

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -2198,6 +2242,28 @@ def __init__(self, shape: pyct.SquareShape):
+    @pycrt.enforce_precision(o=True)
+    def trace(self, **kwargs):
+        """
+        Compute the trace of a squared linear operator.
```
Approximate the trace of a squared linear operator.

joanrue committed:
[pycsou.math][sphinx] add pycsou.math in documentation

joanrue committed:
[pre-commit/black][BugFix] update black version
Related to this known error

joanrue committed:
[lipschitz/hutchpp/trace][DOC] improved documentation

joanrue committed:
[operator/LinOp] don't change class of squared linop to SquareOp
return back to the original init. This was changed in order to have trace() available for squared LinOp. The user should use SquareOp if the intention is to use trace().

joanrue committed:
[operator/LinOp/gram] specialize (co)gram to SelfAdjointOp instead of SquareOp

joanrue committed:
[operator/LinOp/lipschitz] improve default behavior for kwargs
Made more pythonic the removal of "gpu" and added "xp" in case "gpu" is not defined

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1633,8 +1633,6 @@ def __init__(self, shape: pyct.NonAgnosticShape):
         super(LinOp, self).__init__(shape)
         self._diff_lipschitz = 0
-        if shape[0] == shape[1]:
-            self.__class__ = SquareOp
```
Remove the text in docstring related to that feature.

matthieumeo commented on .pre-commit-config.yaml:
```diff
@@ -1,7 +1,7 @@
 repos:
 - repo: https://github.com/psf/black
-  rev: 21.12b0
+  rev: 22.3.0
```
Why the upgrade?

matthieumeo committed:
Merge branch 'v2-dev' into hutchpp
